File: leetcode/algorithms/383_ransom_note/main.py
from collections import Counter


class RansomNote:
    """
    # dict
    #
    # Complexities:
    #   N - The Size of `ransomNote`
    #   M - The Size of `magazine`
    #   K - The Size of Characters in `magazine`
    #   - Time Complexity: O(N + M)
    #   - Space Complexity: O(K)
    """
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        magazineMap: dict = {}

        for c in magazine:
            if magazineMap.get(c) != None:
                magazineMap.update({c: magazineMap.get(c) + 1})
            else:
                magazineMap.update({c: 1})

        for c in ransomNote:
            if magazineMap.get(c) != None and magazineMap.get(c) > 0:
                magazineMap.update({c: magazineMap.get(c) - 1})
            else:
                return False

        return True


    # Solution
    """
    # Solution 1
    #
    # collections.Counter
    #
    # Complexities:
    #   N - The Size of `ransomNote`
    #   M - The Size of `magazine`
    #   - Time Complexity: O(N + M)
    #   - Space Complexity: O(1)
    """
    def solution1(self, ransomNote, magazine):
        return Counter(ransomNote) <= Counter(magazine)

    """
    # Solution 2
    #
    # Counting Array
    #
    # Complexities:
    #   N - The Size of `ransomNote`
    #   M - The Size of `magazine`
    #   - Time Complexity: O(N + M)
    #   - Space Complexity: O(1)
    """
    def solution2(self, ransomNote, magazine):
        if len(ransomNote) > len(magazine):
            return False

        count = [0] * 26
        base = ord('a')

        for c in magazine:
            count[ord(c) - base] += 1

        for c in ransomNote:
            idx = ord(c) - base
            count[idx] -= 1
            if count[idx] < 0:
                return False

        return True

File: leetcode/algorithms/383_ransom_note/main.js
/**
 * Map
 *
 * Complexities:
 *   N - The Size of `ransomNote`
 *   M - The Size of `magazine`
 *   K - The Size of Characters in `magazine`
 *   - Time Complexity: O(N + M)
 *   - Space Complexity: O(K)
 */
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function (ransomNote, magazine) {
  const magazineMap = new Map();

  for (const v of magazine) {
    if (magazineMap.get(v)) {
      magazineMap.set(v, magazineMap.get(v) + 1);
    } else {
      magazineMap.set(v, 1);
    }
  }

  for (const v of ransomNote) {
    if (magazineMap.get(v) && magazineMap.get(v) > 0) {
      magazineMap.set(v, magazineMap.get(v) - 1);
    } else {
      return false;
    }
  }

  return true;
};


// Solution
/**
 * Solution 1
 *
 * Counting Array
 *
 * Complexities:
 *   N - The Size of `ransomNote`
 *   M - The Size of `magazine`
 *   - Time Complexity: O(N + M)
 *   - Space Complexity: O(1)
 */
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var solution1 = function (ransomNote, magazine) {
  if (ransomNote.length > magazine.length) {
    return false;
  }

  const count = new Array(26).fill(0);
  const base = "a".charCodeAt(0);

  for (const c of magazine) {
    count[c.charCodeAt(0) - base]++;
  }

  for (const c of ransomNote) {
    const idx = c.charCodeAt(0) - base;

    if (--count[idx] < 0) {
      return false;
    }
  }

  return true;
};

/**
 * Solution 2
 *
 * Map
 *
 * Complexities:
 *   N - The Size of `ransomNote`
 *   M - The Size of `magazine`
 *   K - The Size of Characters in `magazine`
 *   - Time Complexity: O(N + M)
 *   - Space Complexity: O(K)
 */
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var solution2 = function (ransomNote, magazine) {
  const count = new Map();

  for (const c of magazine) {
    count.set(c, (count.get(c) ?? 0) + 1);
  }

  for (const c of ransomNote) {
    const remaining = count.get(c) ?? 0;

    if (remaining === 0) {
      return false;
    }

    count.set(c, remaining - 1);
  }

  return true;
};
